feat: Revamping Sidebar

JeevaRamanathan opened the pull request:
https://github.com/mfts/papermark/blob/c27f3f60b7e33527eaf2f57b9a92aa6a796a3b19/components/Sidebar.tsx#L156

Hi @mfts,
Since ShadCN introduced the sidebar component two days ago, I thought of revamping the current sidebar using ShadCN's component. I would like to get your input on this.

Here’s a preview of how it looks:
![image](https://github.com/user-attachments/assets/9af293fa-e947-45f5-a25c-58ec7ae96a6e)
Mobile View: 
![image](https://github.com/user-attachments/assets/23b5d037-140a-4d74-b415-501528f9e599)

mfts commented:
@JeevaRamanathan nice. I was hoping someone would add Shadcn sidebar :)

JeevaRamanathan commented:
@mfts  I would like to get your input on this.
Do you prefer keeping the sidebar icons visible when it’s closed, or would you rather have a full-screen version like this example from ShadCN: https://ui.shadcn.com/blocks/default/sidebar-05 and kindly let me know if you have any other inputs or suggestions.

mfts commented:
@JeevaRamanathan 

> Do you prefer keeping the sidebar icons visible when it’s closed

Yes it would be great if the icons would still be visible when closed.

mfts commented:
@JeevaRamanathan let me know when you want me to start reviewing this PR or have questions :)

JeevaRamanathan commented:
Hi @mfts,
Sorry for the delay, was in the middle of something :)

I have removed the commented lines of the previous sidebar along with few changes. I also noticed the visitor option was added, and I have included that as well as resolved the henceforth conflict. 
I can say it's ready for review, but there are two things to note:

1. Currently, the sidebar is not controlled (i.e., when switching pages, it returns to its default open/close state—I'm looking into that).
2. Adding this sidebar creates multiple scroll bars, including one in the AppLayout ([here](https://github.com/JeevaRamanathan/papermark/blob/5b546748966f93132e06b4d829f68cb94bf247d7/components/layouts/app.tsx#L7)). While adding overflow-hidden removes the extra scroll bars, it makes the items partially hidden at the bottom. Your input on this would be helpful.

JeevaRamanathan commented:
Hi @mfts, would you mind having a look into this and suggestions if any based on your time.

JeevaRamanathan commented:
Hi @mfts, would you mind having a look into this and provide suggestions if any based on your time, Thanks :)

JeevaRamanathan commented:
Hi @mfts,
Finally, the sidebar controlled state issue was resolved with help from [this X thread](https://x.com/felixlyu_1018/status/1850201643951996996), and it’s now working locally. However, the changes also involve updates in `ui/sidebar.tsx`. Since a PR might eventually be proposed to ShadCN by the thread author involving changes to `ui/sidebar.tsx`, I made the changes specifically in ui/sidebar.tsx for this PR instead. Can we address this as a separate issue after this PR once the sidebar is updated in ShadCN. 

Your thoughts on this please?

mfts commented on hooks/use-mobile.tsx:
@JeevaRamanathan we already have a function for mobile detection. 

is your use-mobile required for some sidebar stuff? does it come from shadcn or is it your implementation?

https://github.com/mfts/papermark/blob/47802377727118f3e36373d2b90a43040c82fde4/lib/utils/use-media-query.ts

JeevaRamanathan replied:
Yes, the `use-mobile` function was automatically created when installing the sidebar from ShadCN, so it’s part of ShadCN's implementation.

mfts commented on tailwind.config.js:
@JeevaRamanathan could you please add the comments back into their place. these are helpful when working on the tailwind config

JeevaRamanathan replied:
Ok sure!

mfts replied:
much more readable diff! Thanks

mfts reviewed (changes requested):
@JeevaRamanathan great progress. 

I think the collapsible icon should be on the sidebar itself, not on the main div. 

the icons and font of the items are one size too small

On mobile, the navbar should contain the collabsible buttons (which is does) and the user dropdown, which it used to but not longer.

Remove the P icon next to Papermark, because there's no space for the plan name

![CleanShot 2024-10-29 at 12 45 31@2x](https://github.com/user-attachments/assets/8b3d187c-d289-495a-98ea-54ff500e45aa)

![CleanShot 2024-10-29 at 12 48 21@2x](https://github.com/user-attachments/assets/869d07e3-4539-4fe4-9476-f898fa24d281)

JeevaRamanathan commented:
Hi @mfts,

> I think the collapsible icon should be on the sidebar itself, not on the main div.

I felt placing the icon separately from the sidebar might provide more clarity, as the sidebar already contains the title, and combining them could make it feel cluttered. Here’s a similar approach used in ShadCN’s sidebar example, where the trigger is in the main div: [Sidebar example](https://ui.shadcn.com/blocks/default/sidebar-08) and [documentation](https://ui.shadcn.com/docs/components/sidebar#usage).

>the icons and font of the items are one size too small

Got it! I’ll check on this

> and the user dropdown, which it used to but not longer.

I’m not sure I completely understood this part. Could you clarify? Are you suggesting that the user dropdown should be removed from the sidebar on mobile view, possibly moving it to the highlighted rounded area on the right side like now?

>Remove the P icon next to Papermark, because there's no space for the plan name

Understood! I’ll remove the icon when the sidebar is expanded, and I’ll keep it when it’s collapsed to avoid leaving an empty space.

JeevaRamanathan commented:
Hi @mfts, Can you please review the changes
1. Removed the P icon when sidebar is open.
2. Added back the comments in [tailwind.config.js](https://github.com/mfts/papermark/pull/1162/files/2ba3bd3ba80886631e4d3b9ae83eb4fa28e280f7#diff-ddc1067f080930bc48865f4660316a14b21bbc9f9ba85975ce8479342b983d2e)
3. Moved the user dropdown to top in mobile view.
4.The font size remains the same as before, but the icon size was smaller than specified. This appears to be a bug in the shadcn sidebar: [[bug]: Sidebar icon size is not taken in consideration](https://github.com/shadcn-ui/ui/issues/5504).  Implemented a temporary workaround with custom CSS and made a minor change in ui/sidebar.tsx.

Could you please take a look and let me know if any changes are needed?

Thanks!

mfts commented:
@JeevaRamanathan 

> I felt placing the icon separately from the sidebar might provide more clarity, as the sidebar already contains the title, and combining them could make it feel cluttered. Here’s a similar approach used in ShadCN’s sidebar example, where the trigger is in the main div: [Sidebar example](https://ui.shadcn.com/blocks/default/sidebar-08) and [documentation](https://ui.shadcn.com/docs/components/sidebar#usage).

I agree. 
However, I don't love how it's currently solved. We are wasting a lot of space on top. 

I think the https://ui.shadcn.com/blocks#sidebar-07 should be our role model
![CleanShot 2024-10-31 at 13 27 14@2x](https://github.com/user-attachments/assets/2ec8d33c-b3f8-494e-8829-3fa48ebdba7f)

- Let's keep Papermark at the top. 
- move the plan into the dropdown team picker.
- maybe we can even use a nicer dropdown or whatever shadcn block uses here 

![CleanShot 2024-10-31 at 13 28 43@2x](https://github.com/user-attachments/assets/83949bc7-a3cb-4089-a58d-2ff588f4d566)
![CleanShot 2024-10-31 at 13 32 47@2x](https://github.com/user-attachments/assets/0be77942-2dda-41e3-8af0-d924c555c2ef)
![CleanShot 2024-10-31 at 13 33 15@2x](https://github.com/user-attachments/assets/558e3033-1af7-405c-bf2c-2c985b6e78b7)

I also like these collapsibles to reveal more settings 
![CleanShot 2024-10-31 at 13 37 30@2x](https://github.com/user-attachments/assets/181cd91e-0f73-416d-9a8c-f83cda0eacda)

mfts commented:
I'm awarding you now but I'd love for you to continue with this sidebar so we can get it merged :)

/award 750

JeevaRamanathan commented:
Hi @mfts,

> * Let's keep Papermark at the top.
> * move the plan into the dropdown team picker.
> * maybe we can even use a nicer dropdown or whatever shadcn block uses here

I have made these changes similar to sidebar-07. 

> I also like these collapsibles to reveal more settings

We already have branding as a separate option within the settings screen. Should we move the branding section into the collapsible settings as well to reveal more options?
